File: linked_list/easy/reverse_linked_list_206.go
package easy

import "github.com/vielendanke/algorithms_in_place/leetcode/easy"

func reverse(head *easy.ListNode) *easy.ListNode {
	var previousNode, currentNode *easy.ListNode = nil, head

	for currentNode != nil {
		nextNode := currentNode.Next
		currentNode.Next = previousNode
		previousNode = currentNode
		currentNode = nextNode
	}
	return previousNode
}

func reverseRecursive(head, prev *easy.ListNode) *easy.ListNode {
	if head == nil {
		return prev
	}
	next := head.Next
	head.Next = prev
	return reverseRecursive(next, head)
}

File: linked_list/easy/java_solutions/ReverseLinkedList_206.java
package linked_list.easy.java_solutions;

import java.util.ArrayList;
import java.util.List;

import static linked_list.Helper.ListNode;

public class ReverseLinkedList_206 {

    public ListNode reverseListWithArray(ListNode head) {
        List<Integer> list = new ArrayList<>();

        var temp = head;

        while (temp != null) {
            list.add(temp.val);
            temp = temp.next;
        }
        temp = head;
        for (int i = list.size() - 1; i >= 0; i--) {
            temp.val = list.get(i);
            temp = temp.next;
        }
        return head;
    }

    public ListNode reverseList(ListNode head) {
        ListNode prev = null;
        ListNode current = head;

        while (current != null) {
            ListNode nextNode = current.next;
            current.next = prev;
            prev = current;
            current = nextNode;
        }
        return prev;
    }
}
